kotlinx-coroutines-core/native/src/MultithreadedDispatchers.kt

```diff
@@ -73,38 +74,75 @@ private class MultiWorkerDispatcher(
     workersCount: Int
 ) : CloseableCoroutineDispatcher() {
     private val tasksQueue = Channel<Runnable>(Channel.UNLIMITED)
+    private val availableWorkers = Channel<Channel<Runnable>>(Channel.UNLIMITED)
     private val workerPool = OnDemandAllocatingPool(workersCount) {
         Worker.start(name = "$name-$it").apply {
             executeAfter { workerRunLoop() }
         }
     }
 
+    /**
+     * (number of tasks - number of workers) * 2 + (1 if closed)
+     */
+    private val tasksAndWorkersCounter = atomic(0L)
+
+    private inline fun Long.isClosed() = this and 1L == 1L
+    private inline fun Long.hasTasks() = this >= 2
+    private inline fun Long.hasWorkers() = this < 0
+
     private fun workerRunLoop() = runBlocking {
-        // NB: we leverage tail-call optimization in this loop, do not replace it with
-        // .receive() without proper evaluation
-        for (task in tasksQueue) {
-            /**
-             * Any unhandled exception here will pass through worker's boundary and will be properly reported.
-             */
-            task.run()
+        val privateChannel = Channel<Runnable>(1)
+        while (true) {
+            val state = tasksAndWorkersCounter.getAndUpdate {
+                if (it.isClosed() && !it.hasTasks()) return@runBlocking
+                it - 2
+            }
+            if (state.hasTasks()) {
+                // we promised to process a task, and there are some
+                tasksQueue.receive().run()
+            } else {
+                availableWorkers.send(privateChannel)
+                val task = privateChannel.receiveCatching().getOrNull()?.run()
```

Comment:
I think we could quite easily add a `withTimeout` here, then `getAndUpdate { if still considered to be sleeping, remove oneself and terminate }`, and retry the `obtainWorker()`, and this will allow graceful worker termination. I think it would complicate the code somewhat, so didn't initially implement this. @qwwdfsad, do you think this could be worth it?

Reply:
My vote is to have a follow-up PR with the change